src/CarbonAware.LocationSources/src/LocationSource.cs

```diff
@@ -0,0 +1,137 @@
+using CarbonAware.Model;
+using CarbonAware.Interfaces;
+using CarbonAware.Exceptions;
+using Microsoft.Extensions.Logging;
+using System.Text.Json;
+using Microsoft.Extensions.Options;
+using CarbonAware.LocationSources.Configuration;
+
+namespace CarbonAware.LocationSources;
+
+/// <summary>
+/// Represents a location source.
+/// </summary>
+public class LocationSource : ILocationSource
+{
+    private readonly ILogger<LocationSource> _logger;
+
+    private IDictionary<string, NamedGeoposition>? _namedGeopositions;
+
+    private static readonly JsonSerializerOptions options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
+
+    private IOptionsMonitor<LocationDataSourcesConfiguration> _configurationMonitor { get; }
+
+    private LocationDataSourcesConfiguration _configuration => _configurationMonitor.CurrentValue;
+
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="LocationSource"/> class.
+    /// </summary>
+    /// <param name="logger">The logger for the LocationSource</param>
+    public LocationSource(ILogger<LocationSource> logger, IOptionsMonitor<LocationDataSourcesConfiguration> monitor)
+    {
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        _configurationMonitor = monitor ?? throw new ArgumentNullException(nameof(monitor));
+    }
+
+    public async Task<Location> ToGeopositionLocationAsync(Location location)
+    {
+        switch (location.LocationType)
+        {
+            case LocationType.Geoposition:
+            {
+                return await Task.FromResult(location);
+            }
+            case LocationType.CloudProvider: 
+            {
+                var geoPositionLocation = await GetGeoPositionLocationOrThrowAsync(location);
+                return geoPositionLocation;
+            }
+        }
+        throw new LocationConversionException($"Location '{ location.LocationType }' cannot be converted to Geoposition. ");
+    }
+
+    private async Task<Location> GetGeoPositionLocationOrThrowAsync(Location location)
+    {
+        await LoadRegionsFromFileIfNotPresentAsync();
```

Comment:
There are a lot of references to `region`, `regionName` etc.  I think we should just use the term `Name` in this new generic approach.  Not everything is a "region".